### PR TITLE
Coarse-grained debug utilities for SpiNNaker application

### LINKED ISSUE #155: Support reading IO Buffer
This would make debugging pleasant :)

### PR DESCRIPTION
**Pull request by mossblaser**
* [x] Allow reading of the io_printf'd IOBUF data from a SpiNNaker core. (Fixes #155)
* [x] Standalone IOBUF reading CLI app
* [x] List (locations of) cores in a particular state (System-wide 'PS'?)
* [x] ~~Send signals from the CLI(?)~~

Example `rig-iobuf` usage:

    $ rig-iobuf 192.168.1.1 0 0 1
    INFO: Reinjector counters are at address 0x00000000
    INFO: Copied 1836 bytes of commands from SDRAM...
    INFO: Disabling soft_wdog
    INFO: spin1_start
    INFO: Starting main loop with first command at 0x00400dd0

Example `rig-ps` usage to find a misbehaving core (note the Python regex for the state filter):

    $ rig-ps 192.168.1.1 --state '(?!run)' --app-id 66
    X   Y   P   State             Application      App ID
    --- --- --- ----------------- ---------------- ------
      3   6  13 watchdog          network_tester       66

### CODE REVIEW

**Comment by mundya** [comment]
:)

**Comment by mundya** [comment]
Yay :)

**Comment by mossblaser** [comment]
This should be ready to go though travis will fail due to the SpiNN-5 board being on holiday ;).

@mundya and @neworderofjamie The `rig-ps` command will hopefully prove itself very useful! In particular it should help answer questions such as "where did the annealer put my cores!?" and "OK so one core didn't reach a barrier, which one was it and *why*?".

**Comment by mossblaser** [comment]
What a faff... Travis' re-run button doesn't respect the maximum instance count so I've had to re-run each test individually, one after the other... Pfft... Anyway, @mundya; awaiting your review ;).

**Comment by mundya** [comment]
Thank you for this, do you mind if I review tomorrow? Not with it today.

**Comment by mossblaser** [comment]
No problem :) Thanks very much! :)

File: rig/machine_control/machine_controller.py [hunk 1]
```diff
@@ -539,6 +539,28 @@ def get_processor_status(self, p, x, y):
...
+        address = self.read_vcpu_struct_field("iobuf", x, y, p)
+
+        iobuf = ""
+
+        while address:
+            # The IOBUF data is proceeded by a header which gives the next
+            # address and also the length of the string in the current buffer.
+            iobuf_data = self.read(address, iobuf_size + 16, x, y)
+            address, time, ms, length = struct.unpack("<4I", iobuf_data[:16])
+            iobuf += iobuf_data[16:16 + length].decode("utf-8")
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: Would it be more efficient to do this as the following?

```python
iobuf = list()
    # ...
    iobuf.append(...)

iobuf_data = ''.join(iobuf)
```

Comment (mossblaser) [inline rig/machine_control/machine_controller.py]: Quite possibly...

That said, given this is a feature for interactive use I'd argue it probably doesn't matter enough to be worth changing. Do you have strong feelings?

Comment (mundya) [inline rig/machine_control/machine_controller.py]: No :) Just thought I'd ask.

**Comment by mundya** [comment]
This looks good to me :) Thank you!

LGTM

**Comment by mossblaser** [comment]
Thanks!

**Comment by mossblaser** [comment]
(Will bump+release after #171 is merged)

**Pull request merged**